workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,228 @@
+        private readonly Entity entity;
+        private readonly long spatialId;
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> behavioursRequiringReaderTypes
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> behavioursRequiringWriterTypes
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<MonoBehaviour, int> numUnsatisfiedReadersOrWriters
+            = new Dictionary<MonoBehaviour, int>();
+
+        private readonly Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>> behaviourToReadersWriters
+            = new Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>>();
+
+        private readonly Dictionary<uint, HashSet<IReaderInternal>> compIdToReadersWriters =
+            new Dictionary<uint, HashSet<IReaderInternal>>();
 
-        public SpatialOSBehaviourManager(GameObject gameObject)
+        private readonly HashSet<MonoBehaviour> behavioursToEnable = new HashSet<MonoBehaviour>();
+        private readonly HashSet<MonoBehaviour> behavioursToDisable = new HashSet<MonoBehaviour>();
+
+        private readonly SpatialOSBehaviourLibrary behaviourLibrary;
+
+        private readonly ILogDispatcher logger;
+
+        private const string LoggerName = "SpatialOSBehaviourManager";
+
+        public SpatialOSBehaviourManager(GameObject gameObject, SpatialOSBehaviourLibrary library, ILogDispatcher logger)
         {
-            // Setup caches
+            this.logger = logger;
+            behaviourLibrary = library;
+
+            var spatialComponent = gameObject.GetComponent<SpatialOSComponent>();
+            entity = spatialComponent.Entity;
+            spatialId = spatialComponent.SpatialEntityId;
+
+            foreach (var behaviour in gameObject.GetComponents<MonoBehaviour>())
+            {
+                var readerIds = library.GetRequiredReaderComponentIds(behaviour.GetType());
+                foreach (var id in readerIds)
+                {
+                    if (!behavioursRequiringReaderTypes.TryGetValue(id, out var behavioursForReader))
+                    {
+                        behavioursForReader = new HashSet<MonoBehaviour>();
+                        behavioursRequiringReaderTypes[id] = behavioursForReader;
+                    }
+
+                    behavioursForReader.Add(behaviour);
+                }
+
+                var writerIds = library.GetRequiredWriterComponentIds(behaviour.GetType());
+                foreach (var id in writerIds)
+                {
+                    if (!behavioursRequiringWriterTypes.TryGetValue(id, out var behavioursForWriter))
+                    {
+                        behavioursForWriter = new HashSet<MonoBehaviour>();
+                        behavioursRequiringWriterTypes[id] = behavioursForWriter;
+                    }
```

Comment:
This pattern shows up a few times - you can turn this into a generic helper method I think: 

```
        private TValue GetOrCreateValue<TKey, TValue>(Dictionary<TKey, TValue> dictionary, TKey key) where TValue : new()
        {
            if (!dictionary.TryGetValue(key, out var value))
            {
                value = new TValue();
                dictionary[key] = value;
            }

            return value;
        }
```

Reply:
Done